### ISSUE TITLE
Memory leak when replication targets are unhealthy 

### ISSUE THREAD

**Issue by pbedat** [benbjohnson/litestream#562]
I have the following replication setup:

We are replicating ~40 databases to two cifs mounts (Storageboxes hosted by Hetzner). Those boxes are sometimes undergoing maintenance and one of the mounts go bad.
This happened on 27.01. around 11:15 AM and from this point memory steadily increased with each snapshot interval (4h):

![Screenshot from 2024-01-28 19-57-58](https://github.com/benbjohnson/litestream/assets/1275613/52d608dd-f732-485d-b609-269f6ccad798)

I'm also seeing the following error logs:

```
"snapshots: cannot fetch generations: open /opt/copilot/data/replicas-2/milchsackfabrik/generations: permission denied"
```

It's not a huge problem, but I just wanted to report it.
If you need anything, I could setup a replication environment and take readings.

Edit: v0.3.13

**Comment by hifi** [comment]
It's likely the LZ4 compression library as it never frees a buffer pool it keeps and error conditions seem to hit it hard. Though this pattern is quite suspicious as I'd expect it to be able to reuse the same pool so it might not be it. 

It you can take pprof memory dump from a repro it would answer the question where the memory is actually going to.

**Comment by pbedat** [comment]
@hifi I guess you are right about lz4.

![profile001](https://github.com/benbjohnson/litestream/assets/1275613/6c1c2654-d08e-40a5-9b03-6bb36e6d9611)

[pprof.litestream.alloc_objects.alloc_space.inuse_objects.inuse_space.001.pb.gz](https://github.com/benbjohnson/litestream/files/14077467/pprof.litestream.alloc_objects.alloc_space.inuse_objects.inuse_space.001.pb.gz)

PS: I took the snapshot from replication setup with only 4 DBs. Hence the smaller size.

**Comment by hifi** [comment]
Yeah, unfortunately there's nothing Litestream can do except change implementation or to support multiple compression schemes with different tradeoffs like CPU over RAM.

There's an open issue against the lz4 library about the ever growing pool but the author hadn't responded to it last time I checked.

**Comment by pbedat** [comment]
Never mind. It's not a problem, since I'm getting alerts, when replicas go unhealthy and can respond before memory runs out.
Thank's for clearifying it!
